NameLayerMain/src/vg/civcraft/mc/namelayer/gui/MainGroupGUI.java

```diff
@@ -0,0 +1,1272 @@
+					mod,
+					ChatColor.GOLD
+							+ demoteOrPromote(g.getPlayerType(toChange), pType,
+									true) + " this player to "
+							+ getDirectRankName(pType));
+			if (!gm.hasAccess(g, p.getUniqueId(), getAccordingPermission(pType))) {
+				ISUtils.addLore(mod, ChatColor.RED
+						+ "You dont have permission to do this");
+				modClick = new DecorationStack(mod);
+			} else {
+				modClick = new Clickable(mod) {
+
+					@Override
+					public void clicked(Player arg0) {
+						changePlayerRank(toChange, pType);
+						showDetail(toChange);
+					}
+				};
+			}
+		}
+		return modClick;
+	}
+
+	private void removeMember(UUID toRemove) {
+		if (gm.hasAccess(g, p.getUniqueId(),
+				getAccordingPermission(g.getCurrentRank(toRemove)))) {
+			if (!g.isMember(toRemove)) {
+				p.sendMessage(ChatColor.RED
+						+ "This player is no longer on the group and can't be removed");
+				return;
+			}
+			if (g.isOwner(toRemove)) {
+				p.sendMessage(ChatColor.RED
+						+ "This player owns the group and can't be removed");
+			}
+			NameLayerPlugin.log(Level.INFO,
+					p.getName() + " kicked " + NameAPI.getCurrentName(toRemove)
+							+ " from " + g.getName() + "via gui");
+			g.removeMember(toRemove);
+			checkRecacheGroup();
+			p.sendMessage(ChatColor.GREEN + NameAPI.getCurrentName(toRemove)
+					+ " has been removed from the group");
+		} else {
+			p.sendMessage(ChatColor.RED
+					+ "You have lost permission to remove this player");
+		}
+	}
+
+	private void changePlayerRank(UUID toChange, PlayerType newRank) {
+		if (gm.hasAccess(g, p.getUniqueId(),
+				getAccordingPermission(g.getCurrentRank(toChange)))) {
+			if (!g.isMember(toChange)) {
+				p.sendMessage(ChatColor.RED
+						+ "This player is no longer on the group and can't be "
+						+ demoteOrPromote(g.getCurrentRank(toChange), newRank,
+								false) + "d");
+				return;
+			}
+			if (g.isOwner(toChange)) {
+				p.sendMessage(ChatColor.RED
+						+ "This player owns the group and can't be demoted");
+			}
+			OfflinePlayer prom = Bukkit.getOfflinePlayer(toChange);
+			NameLayerPlugin.log(
+					Level.INFO,
+					p.getName() + " changed player rank for "
+							+ NameAPI.getCurrentName(toChange) + " from "
+							+ g.getCurrentRank(toChange).toString() + " to "
+							+ newRank.toString() + " for group " + g.getName()
+							+ "via gui");
+			if (prom.isOnline()) {
+				Player oProm = (Player) prom;
+				PromotePlayerEvent event = new PromotePlayerEvent(oProm, g,
+						g.getCurrentRank(toChange), newRank);
+				Bukkit.getPluginManager().callEvent(event);
+				if (event.isCancelled()) {
+					p.sendMessage(ChatColor.RED
+							+ "Could not change player rank, you should complain about this");
+					return;
+				}
+				g.removeMember(toChange);
+				g.addMember(toChange, newRank);
+				oProm.sendMessage(ChatColor.GREEN
+						+ "You have been promoted to " + getRankName(toChange)
+						+ " in (Group) " + g.getName());
+			} else {
+				// player is offline change their perms
```

Comment:
Wonder if its worth a TODO here to inject an OfflinePromotePlayerEvent passing in the OfflinePlayer object instead of Player object.

Thoughts? I don't know that anything is using it, but making the event system as easy to bypass as being offline seems wrong.